### PR TITLE
refactor: `ensurePr` improved testability

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Mock every function external to `ensurePr`
- Fix test for `schedule.ts`

## Context

- Ref: #15276

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/repository/update/pr/index.ts [hunk 1]
```diff
@@ -318,17 +316,15 @@ export async function ensurePr(
...
-      // istanbul ignore if
-      if (GlobalConfig.get('dryRun')) {
-        logger.info('DRY-RUN: Would create PR: ' + prTitle);
-        pr = { number: 0, displayNumber: 'Dry run PR' } as never;
-      } else {
+    if (GlobalConfig.get('dryRun')) {
+      logger.info('DRY-RUN: Would create PR: ' + prTitle);
+      pr = { number: 0, displayNumber: 'Dry run PR' } as never;
+    } else {
+      try {
```
Comment (zharinov) [inline lib/workers/repository/update/pr/index.ts:327]:
```ts
try {
  if (A) {
    B
  } else {
    C
  }
} catch <...>
```

becomes

```ts
if (A) {
  B
} else {
  try {
    C
  } catch <...>
}
```

It eliminates code which isn't achievable

File: lib/workers/repository/update/pr/index.ts [hunk 2]
```diff
@@ -348,31 +344,27 @@ export async function ensurePr(
...
-        logger.warn('A pull requests already exists');
-        return { type: 'without-pr', prBlockedBy: 'Error' };
-      }
-      if (err.statusCode === 502) {
-        logger.warn(
-          { branch: branchName },
-          'Deleting branch due to server error'
-        );
-        if (GlobalConfig.get('dryRun')) {
-          logger.info('DRY-RUN: Would delete branch: ' + config.branchName);
```
Comment (rarkins) [inline lib/workers/repository/update/pr/index.ts:370]: When doing a "hide whitespace" review of this PR, I don't see where this if clause was relocated to, so it looks like it's removed. Please double check and confirm that it's intentional

File: lib/workers/repository/update/pr/index.ts [hunk 3]
```diff
@@ -412,9 +403,9 @@ export async function ensurePr(
     logger.debug(`Created ${pr.displayNumber}`);
     return { type: 'with-pr', pr };
   } catch (err) {
-    // istanbul ignore if
+    const x = err instanceof ExternalHostError;
     if (
-      err instanceof ExternalHostError ||
+      x ||
```
Comment (rarkins) [inline lib/workers/repository/update/pr/index.ts:408]: Why this change?

File: lib/workers/repository/update/pr/index.ts (hunk 2 above)
Comment (zharinov) [inline lib/workers/repository/update/pr/index.ts:370]: This isn't achievable

File: lib/workers/repository/update/pr/index.ts (hunk 1 above)
Comment (zharinov) [reply lib/workers/repository/update/pr/index.ts:327]: @rarkins Here is the reason for the only non-comment modification in `index.ts`

**Review by rarkins — DISMISSED** [review]

File: lib/workers/repository/update/pr/index.ts (hunk 2 above)
Comment (zharinov) [reply lib/workers/repository/update/pr/index.ts:370]: image

**Review by viceice — COMMENTED** [review]
did you validated that at least the modified files are strict null safe? Can you fix those without big modifications?

**Review by rarkins — DISMISSED** [review]

**Review by rarkins — DISMISSED** [review]

**Comment by zharinov** [comment]
Took a while to go this extra mile

**Review by rarkins — APPROVED** [review]
